daffodil-core/src/main/scala/org/apache/daffodil/core/dpath/Expression.scala

```diff
@@ -2295,6 +2295,18 @@ abstract class FunctionCallBase(
     res
   }
 
+  final def checkArgArray(): Unit = {
+    lazy val isArray = expressions.head match {
+      case rpe: RelativePathExpression => rpe.steps.last.isArray
+      case rpe: RootPathExpression => rpe.steps.last.isArray
+      case _ => true
```

Comment:
One thing to consider is this doesn't catch expressions like this:
```xquery
fn:count(if (some_condition) then /some/path else /some/other/path)
```
I think that's legal and I think would be considered invalid because of this case.

I'm not sure if there's a good way to check that all branches of a subexpression result in an array. We must be able to get some basic information since we can statically check that all branches have the same type. I think we do that via `inherentType`, but I'm not sure the result would return information about being an array or not. Maybe this is just a limitation of our checking and it's good enough? And if someone whats the above, they would need to do something like this:
```xquery
if (some_condition) then fn:count(/some/path) else fn:count(/some/other/path)
```

Reply:
> One thing to consider is this doesn't catch expressions like this:
> 
> ```xq
> fn:count(if (some_condition) then /some/path else /some/other/path)
> ```
> 
> I think that's legal and I think would be considered invalid because of this case.
> 
> I'm not sure if there's a good way to check that all branches of a subexpression result in an array. We must be able to get some basic information since we can statically check that all branches have the same type. I think we do that via `inherentType`, but I'm not sure the result would return information about being an array or not. Maybe this is just a limitation of our checking and it's good enough? And if someone whats the above, they would need to do something like this:
> 
> ```xq
> if (some_condition) then fn:count(/some/path) else fn:count(/some/other/path)
> ```

Pretty sure the expression type system treats "array or optional" as a type, and so the type inference on the expressions requires all branches of an IF to evaluate to an array/optional in order to satisfy the requirements of fn:count. It's supposed to propagate the requirement from the fn:count into the expression, which would then propagate it to the then and else branch expressions as the "target type" that the expression is trying to create.

Reply:
I don't think my original comment is addressed? Or are you saying there is a check somewhere else within the type system that makes sure things work as expected? Are there tests for this?

Reply:
Best to add a test to check so we can know for sure.

Reply:
There are two new tests using subexpressions containing optional and scalar elements.

Reply:
An if-then-else propagates the type needed by the surrounding context of the if-then-else and repeats it for the 'then' and 'else' subexpressions:
```  
override def targetTypeForSubexpression(subexp: Expression) = {
    Assert.invariant(children.contains(subexp))
    if (subexp == predicate)
      NodeInfo.Boolean
    else
      this.targetType
  }
```
But I looked at every occurrence of "count(" in the code base and none of them contain an if-then-else, so we need to add a test.